nucypher/config/storages.py

```diff
@@ -0,0 +1,307 @@
+import binascii
+import os
+import tempfile
+from abc import abstractmethod, ABC
+from logging import getLogger
+
+import boto3 as boto3
+import shutil
+from botocore.errorfactory import ClientError
+from constant_sorrow import constants
+from typing import Callable
+
+from nucypher.config.constants import DEFAULT_CONFIG_ROOT
+
+
+class NodeStorage(ABC):
+
+    _name = NotImplemented
+    _TYPE_LABEL = 'storage_type'
+    NODE_SERIALIZER = binascii.hexlify
+    NODE_DESERIALIZER = binascii.unhexlify
+
+    class NodeStorageError(Exception):
+        pass
+
+    class UnknownNode(NodeStorageError):
+        pass
+
+    def __init__(self,
+                 character_class,
+                 federated_only: bool,  # TODO# 466
+                 serializer: Callable = NODE_SERIALIZER,
+                 deserializer: Callable = NODE_DESERIALIZER,
+                 ) -> None:
+
+        self.log = getLogger(self.__class__.__name__)
+        self.serializer = serializer
+        self.deserializer = deserializer
+        self.federated_only = federated_only
+        self.character_class = character_class
+
+    def __getitem__(self, item):
+        return self.get(checksum_address=item, federated_only=self.federated_only)
+
+    def __setitem__(self, key, value):
+        return self.save(node=value)
+
+    def __delitem__(self, key):
+        self.remove(checksum_address=key)
+
+    def __iter__(self):
+        return self.all(federated_only=self.federated_only)
+
+    @abstractmethod
+    def all(self, federated_only: bool) -> set:
+        raise NotImplementedError
+
+    @abstractmethod
+    def get(self, checksum_address: str, federated_only: bool):
+        raise NotImplementedError
+
+    @abstractmethod
+    def save(self, node):
+        raise NotImplementedError
+
+    @abstractmethod
+    def remove(self, checksum_address: str) -> bool:
+        raise NotImplementedError
+
+    @abstractmethod
+    def payload(self) -> str:
+        raise NotImplementedError
+
+    @classmethod
+    @abstractmethod
+    def from_payload(self, data: str, *args, **kwargs) -> 'NodeStorage':
+        raise NotImplementedError
+
+    @abstractmethod
+    def initialize(self):
+        raise NotImplementedError
+
+
+class InMemoryNodeStorage(NodeStorage):
```

Comment:
Hmm, these are neat.

Reply:
Yes - They help speed up tests and is used when running a node in development mode as the default node storage.